File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/snapshot/SnapshotterManager.kt
package com.mapbox.maps.mapbox_maps.snapshot

import com.mapbox.maps.Size
import com.mapbox.maps.mapbox_maps.toCameraOptions
import com.mapbox.maps.mapbox_maps.toCameraState
import com.mapbox.maps.mapbox_maps.toEdgeInsets
import com.mapbox.maps.mapbox_maps.toFLTCameraOptions
import com.mapbox.maps.mapbox_maps.toFLTCoordinateBounds
import com.mapbox.maps.mapbox_maps.toFLTSize
import com.mapbox.maps.mapbox_maps.toMbxImage
import com.mapbox.maps.mapbox_maps.toPoint
import com.mapbox.maps.pigeons.FLTMapInterfaces
import com.mapbox.maps.pigeons.FLTMapInterfaces.MbxImage
import com.mapbox.maps.pigeons.FLTSnapshot
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CountDownLatch
import java.util.concurrent.CyclicBarrier

class SnapshotterManager(private val delegate: SnapshotControllerDelegate) :
  FLTSnapshot._SnapshotterMessager {
  override fun cancel(id: String) {
    delegate.getSnapshotter(id).cancel();
  }

  override fun destroy(id: String) {
    delegate.getSnapshotter(id).destroy()
    delegate.remove(id)
  }

  override fun setCamera(id: String, cameraOptions: FLTMapInterfaces.CameraOptions) {
    delegate.getSnapshotter(id).setCamera(cameraOptions.toCameraOptions(delegate.getContext()))
  }

  override fun setStyleUri(id: String, styleUri: String) {
    delegate.getSnapshotter(id).setStyleUri(styleUri)
  }

  override fun setStyleJson(id: String, styleJson: String) {
    delegate.getSnapshotter(id).setStyleJson(styleJson)
  }

  override fun setSize(id: String, size: FLTMapInterfaces.Size) {
    delegate.getSnapshotter(id).setSize(Size(size.width.toFloat(), size.height.toFloat()))
  }

  override fun cameraForCoordinates(
    id: String,
    coordinates: MutableList<MutableMap<String, Any>>,
    padding: FLTMapInterfaces.MbxEdgeInsets,
    bearing: Double?,
    pitch: Double?
  ): FLTMapInterfaces.CameraOptions {
    val cameraOptions = delegate.getSnapshotter(id).cameraForCoordinates(
      coordinates.map { it.toPoint() },
      padding.toEdgeInsets(delegate.getContext()),
      bearing ?: 0.0, pitch ?: 0.0,
    )
    return cameraOptions.toFLTCameraOptions(delegate.getContext())
  }

  override fun coordinateBoundsForCamera(
    id: String,
    camera: FLTMapInterfaces.CameraOptions
  ): FLTMapInterfaces.CoordinateBounds {
    val coordinateBounds = delegate.getSnapshotter(id)
      .coordinateBoundsForCamera(camera.toCameraOptions(delegate.getContext()))
    return coordinateBounds.toFLTCoordinateBounds()
  }

  override fun getCameraState(id: String): FLTMapInterfaces.CameraState {
    val cameraState = delegate.getSnapshotter(id).getCameraState()
    return cameraState.toCameraState(delegate.getContext())
  }

  override fun getSize(id: String): FLTMapInterfaces.Size {
    val size = delegate.getSnapshotter(id).getSize()
    return size.toFLTSize(delegate.getContext())
  }

  override fun getStyleJson(id: String): String {
    return delegate.getSnapshotter(id).getStyleJson()
  }

  override fun getStyleUri(id: String): String {
    return delegate.getSnapshotter(id).getStyleUri()
  }

  override fun start(id: String, result: FLTSnapshot.NullableResult<MbxImage>) {
    delegate.getSnapshotter(id).start { snapshot, errorMessage ->
      if (errorMessage == null) {
        result.success(snapshot?.toMbxImage())
      } else {
        result.error(UnknownError(errorMessage))
      }
    }
  }
}

File: ios/Classes/SnapshotterManager.swift
import Foundation
import MapboxMaps

class SnapshotterManager : NSObject, FLT_SnapshotterMessager {
    
    private var delegate: SnapshotControllerDelegate
    private static let errorCode = "0"
    
    init(withDelegate delegate: SnapshotControllerDelegate) {
        self.delegate = delegate
    }
    
    func cancelId(_ id: String, error flutterError: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        do {
            try delegate.getSnapshotter(id: id).cancel()
        } catch {
            flutterError.pointee = FlutterError(code: SnapshotterManager.errorCode, message: error.localizedDescription, details: nil)
        }
    }
    
    func destroyId(_ id: String, error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        delegate.remove(id: id)
    }
    
    func setCameraId(_ id: String, cameraOptions: FLTCameraOptions, error flutterError: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        do {
            try delegate.getSnapshotter(id: id).setCamera(to: cameraOptions.toCameraOptions())
        } catch {
            flutterError.pointee = FlutterError(code: SnapshotterManager.errorCode,message: error.localizedDescription,details: nil)
        }
    }
    
    func setStyleUriId(_ id: String, styleUri: String, error flutterError: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        do {
            try delegate.getSnapshotter(id: id).styleURI = StyleURI.init(rawValue: styleUri)
        } catch {
            flutterError.pointee = FlutterError(code: SnapshotterManager.errorCode,message: error.localizedDescription,details: nil)
        }
    }
    
    func setStyleJsonId(_ id: String, styleJson: String, error flutterError: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        do {
            try delegate.getSnapshotter(id: id).styleJSON = styleJson
        } catch {
            flutterError.pointee = FlutterError(code: SnapshotterManager.errorCode,message: error.localizedDescription,details: nil)
        }
    }
    
    func setSizeId(_ id: String, size: FLTSize, error flutterError: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        do {
            try delegate.getSnapshotter(id: id).snapshotSize = CGSize(width: size.width, height: size.height)
        } catch {
            flutterError.pointee = FlutterError(code: SnapshotterManager.errorCode,message: error.localizedDescription,details: nil)
        }
    }
    
    func camera(forCoordinatesId id: String, coordinates: [[String : Any]], padding: FLTMbxEdgeInsets, bearing: NSNumber?, pitch: NSNumber?, error flutterError: AutoreleasingUnsafeMutablePointer<FlutterError?>) -> FLTCameraOptions? {
        do {
            let cameraOptions = try delegate.getSnapshotter(id: id).camera(for: coordinates.compactMap(convertDictionaryToCLLocationCoordinate2D(dict:)), padding: padding.toUIEdgeInsets(), bearing: bearing?.doubleValue, pitch: pitch?.doubleValue)
            return cameraOptions.toFLTCameraOptions()
        } catch {
            flutterError.pointee = FlutterError(code: SnapshotterManager.errorCode,message: error.localizedDescription,details: nil)
            return nil
        }
    }
    
    func coordinateBounds(forCameraId id: String, camera: FLTCameraOptions, error flutterError: AutoreleasingUnsafeMutablePointer<FlutterError?>) -> FLTCoordinateBounds? {
        do {
            let coordinateBounds = try delegate.getSnapshotter(id: id).coordinateBounds(for: camera.toCameraOptions())
            return coordinateBounds.toFLTCoordinateBounds()
        } catch {
            flutterError.pointee = FlutterError(code: SnapshotterManager.errorCode,message: error.localizedDescription,details: nil)
            return nil
        }
    }
    
    func getCameraStateId(_ id: String, error flutterError: AutoreleasingUnsafeMutablePointer<FlutterError?>) -> FLTCameraState? {
        do {
            let camera = try delegate.getSnapshotter(id: id).cameraState
            return camera.toFLTCameraState()
        } catch {
            flutterError.pointee = FlutterError(code: SnapshotterManager.errorCode,message: error.localizedDescription,details: nil)
            return nil
        }
    }
    
    func getSizeId(_ id: String, error flutterError: AutoreleasingUnsafeMutablePointer<FlutterError?>) -> FLTSize? {
        do {
            let size = try delegate.getSnapshotter(id: id).snapshotSize
            return size.toFLTSize()
        } catch {
            flutterError.pointee = FlutterError(code: SnapshotterManager.errorCode,message: error.localizedDescription,details: nil)
            return nil
        }
    }
    
    func getStyleJsonId(_ id: String, error flutterError: AutoreleasingUnsafeMutablePointer<FlutterError?>) -> String? {
        do {
            let json = try delegate.getSnapshotter(id: id).styleJSON
            return json
        } catch {
            flutterError.pointee = FlutterError(code: SnapshotterManager.errorCode,message: error.localizedDescription,details: nil)
            return nil
        }
    }
    
    func getStyleUriId(_ id: String, error flutterError: AutoreleasingUnsafeMutablePointer<FlutterError?>) -> String? {
        do {
            let uri = try delegate.getSnapshotter(id: id).styleURI
            return uri?.rawValue
        } catch {
            flutterError.pointee = FlutterError(code: SnapshotterManager.errorCode,message: error.localizedDescription,details: nil)
            return nil
        }
    }
    
    func startId(_ id: String, completion: @escaping (FLTMbxImage?, FlutterError?) -> Void) {
        do {
            try delegate.getSnapshotter(id: id).start(overlayHandler: nil) { snapshotResult in
                let image = try? snapshotResult.get()
                if  image != nil {
                    completion(image!.toFLTMbxImage(),nil)
                } else {
                    completion(nil,FlutterError(code: SnapshotterManager.errorCode,message: "snapshot fail",details: nil))
                }
            }
        } catch {
            completion(nil,FlutterError(code: SnapshotterManager.errorCode,message: error.localizedDescription,details: nil))
        }
    }
    
}
